### PR TITLE
Ir specialisation

### PR DESCRIPTION
**Pull request by PalumboN**
The goal of this is to make some small experiments with specialisations in the IR.

A specialisation adds some checks at the beginning of the IR to clean checks inside the primitive (or bytecode).

- Renaming the term `primitive customisation` to `primitive specialisation`
  - Specialisations define how to specialise the IR adding checks at the beginning
  - They can also add more constraints to the IR (for hard equivalents checks, as SmallIntegers by method tags).
- Adding an `oop` constraint, to manage `load` operations constraints.
- Adding some algebra for constraints.

### What is missing (open issues)
- Value numbering to avoid replicated operations -> share constraints https://github.com/Alamvic/druid/issues/78
- A better constraint to say: _"The format of this oop is N"_ https://github.com/Alamvic/druid/issues/79

### CODE REVIEW

File: Druid-Tests/DRIRTest.class.st [hunk 1]
```diff
@@ -41,10 +41,27 @@ DRIRTest >> assertConstraint: aConstraint includes: includedConstraint [
 	self assert: (aConstraint includes: includedConstraint)
 ]
 
+{ #category : #asserting }
+DRIRTest >> assertConstraintIn: aDRCPSEdge for: aKey includes: expectedConstraint [
+
+	| constaint |
+	constaint := aDRCPSEdge constraintFor: aKey.
```
Comment (guillep) [inline Druid-Tests/DRIRTest.class.st:48]: what is a key? a (register) name?

File: Druid-Tests/DRSpecialisationCompilerTest.class.st [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+		for: 'R3'
+		includes: (DREqualsConstraint withValue: 2)
+]
+
+{ #category : #tests }
+DRSpecialisationCompilerTest >> testSmallIntegerConditionalCheck [
+
+	| cfg specialisationCheck |
+	cfg := self generateDruidIRFor: #primitiveAdd.
+	specialisationCheck := cfg allConditionalJumps first.
```
Comment (guillep) [inline Druid-Tests/DRSpecialisationCompilerTest.class.st:59]: here there is no specialization?

File: Druid/DRAbstractCompilerCompiler.class.st [hunk 3]
```diff
@@ -133,7 +131,7 @@ DRAbstractCompilerCompiler >> initialize [
 
 	codeGenerator := DRCogitCodeGenerator new.
 	registerAllocator := DRCogitLinearScanRegisterAllocator new.
-	customisation := DRNonePrimitiveCustomisation new.
+	specialisation := DRNonePrimitiveSpecialisation new.
```
Comment (guillep) [inline Druid/DRAbstractCompilerCompiler.class.st:134]: :)

File: Druid/DRArrayPrimitiveSpecialisation.class.st [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+DRArrayPrimitiveSpecialisation class >> pragmaName [
+
+	^ #arrayOop
+]
+
+{ #category : #api }
+DRArrayPrimitiveSpecialisation >> addInitialConstraint: aDRPrimitiveControlFlowGraph [
+	"Do nothing, let the inference to the added checks"
+
+	"aDRPrimitiveControlFlowGraph initialConstraint: self initialConstraint"
```
Comment (guillep) [inline Druid/DRArrayPrimitiveSpecialisation.class.st:17]: Maybe this method has to go away?

File: Druid/DRArrayPrimitiveSpecialisation.class.st [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+	"if false, fail"
+	conditionalJump falseBranch: aDRPrimitiveIRGenerator ir failureExitBlock.
+	aDRPrimitiveIRGenerator ir failureExitBlock addPredecessor: conditionalBasicBlock.
+
+	"if true, continue"
+	conditionalJump trueBranch: aDRPrimitiveIRGenerator newBasicBlock.
+	aDRPrimitiveIRGenerator currentBasicBlock addPredecessor: conditionalBasicBlock.
+
+	aDRPrimitiveIRGenerator popFrameMergingDeferredReturns
+]
```
Comment (guillep) [inline Druid/DRArrayPrimitiveSpecialisation.class.st:110]: Auch, we need a better way to write this :)

File: Druid/DRCPSEdge.class.st [hunk 6]
```diff
@@ -185,11 +191,21 @@ DRCPSEdge >> visitAnd: aDRAnd [
...
 ]
 
+{ #category : #visiting }
+DRCPSEdge >> visitBitAnd: aDRBitAnd [
+
+	^ (self operandConstraint: aDRBitAnd operand1) ifNotNil: [ :operandConstraint |
+		  aDRBitAnd operand2 isConstant ifTrue: [
+			  (operandConstraint bitAndConstraint: aDRBitAnd operand2 value) ifNotNil: [
+				  :constraint | self addConstraint: constraint to: aDRBitAnd result name ] ] ]
+]
```
Comment (guillep) [inline Druid/DRCPSEdge.class.st:207]: Rough

File: Druid/DRCPSEdge.class.st [hunk 7]
```diff
@@ -426,9 +448,11 @@ DRCPSEdge >> visitLessThan: aDRLessThanComparison inBranch: aDRBranch [
 ]
 
 { #category : #visiting }
-DRCPSEdge >> visitLoad: aDRLoad [ 
+DRCPSEdge >> visitLoad: aDRLoad [
 
-	^ nil
+	^ (aDRLoad address acceptVisitor: self) ifNotNil: [ :addressConstraint |
+		  (addressConstraint loadConstraint: aDRLoad size) ifNotNil: [ :constraint |
+			  self addConstraint: constraint to: aDRLoad result name ] ]
```
Comment (guillep) [inline Druid/DRCPSEdge.class.st:455]: rough

File: Druid/DRCogitStackToRegisterMappingGenerator.class.st [hunk 8]
```diff
@@ -83,41 +82,6 @@ DRCogitStackToRegisterMappingGenerator >> initialize [
 	markDeadCode := false
 ]
 
-{ #category : #'code-generation' }
-DRCogitStackToRegisterMappingGenerator >> innerCompilePathFrom: aDRBasicBlock [
```
Comment (guillep) [inline Druid/DRCogitStackToRegisterMappingGenerator.class.st:87]: Ok this was moved up

File: Druid-Tests/DRIRTest.class.st [hunk 9]
```diff
@@ -58,6 +75,7 @@ DRIRTest >> generateBasicDruidIRFor: methodSelector [
 
 	| method cfg |
 	method := DruidTestInterpreter lookupSelector: methodSelector.
+	compilerCompiler customiseFor: method.
```
Comment (guillep) [inline Druid-Tests/DRIRTest.class.st:78]: ?

File: Druid/DRSmallIntegerPrimitiveSpecialisation.class.st [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+{ #category : #accessing }
+DRSmallIntegerPrimitiveSpecialisation class >> pragmaName [
+
+	^ #smallInteger
+]
+
+{ #category : #api }
+DRSmallIntegerPrimitiveSpecialisation >> addInitialConstraint: anIR [
+
+	anIR initialConstraint: self initialConstraint
```
Comment (guillep) [inline Druid/DRSmallIntegerPrimitiveSpecialisation.class.st:16]: I wonder why do we need two hooks

File: Druid/DRSCCP.class.st [hunk 11]
```diff
@@ -200,9 +200,9 @@ DRSCCP >> latticeValue: aValue [
 DRSCCP >> latticeValueOfPhiFunction: operation [
 
 	| values |
-
-	values := operation operands collect: [ :e | 
-		self latticeValue: e ] as: Set.
+	values := (operation operands
+		           reject: [ :e | e isNullValue ] "Avoid NULLs"
```
Comment (guillep) [inline Druid/DRSCCP.class.st:204]: Can we have nulls???

**Review by guillep — DISMISSED** [review]
Some comments

File: Druid/DRSCCP.class.st (hunk 11 above)
Comment (PalumboN) [reply Druid/DRSCCP.class.st:204]: Yes, I think it comes from messages like `ifNotNil`

File: Druid/DRSmallIntegerPrimitiveSpecialisation.class.st (hunk 10 above)
Comment (PalumboN) [reply Druid/DRSmallIntegerPrimitiveSpecialisation.class.st:16]: What so you refer with "two hooks"?

File: Druid/DRCPSEdge.class.st (hunk 6 above)
Comment (PalumboN) [reply Druid/DRCPSEdge.class.st:207]: First iteration: make it works.

File: Druid/DRCPSEdge.class.st (hunk 7 above)
Comment (PalumboN) [reply Druid/DRCPSEdge.class.st:455]: First iteration: make it works.

File: Druid/DRCPSEdge.class.st (hunk 6 above)
Comment (PalumboN) [reply Druid/DRCPSEdge.class.st:207]: Probably it will not be necessary using PRE

File: Druid/DRArrayPrimitiveSpecialisation.class.st (hunk 5 above)
Comment (PalumboN) [reply Druid/DRArrayPrimitiveSpecialisation.class.st:110]: Yes, probably interpreting Pharo code directly

File: Druid-Tests/DRSpecialisationCompilerTest.class.st (hunk 2 above)
Comment (PalumboN) [reply Druid-Tests/DRSpecialisationCompilerTest.class.st:59]: Yes, it is as a pragma at `primitiveAdd`.

But should be nice do it explicitly in the test 🤓

File: Druid-Tests/DRIRTest.class.st (hunk 1 above)
Comment (PalumboN) [reply Druid-Tests/DRIRTest.class.st:48]: Yes, the string. For now, it was enough.

Comment (guillep) [reply Druid-Tests/DRIRTest.class.st:48]: ok!

File: Druid-Tests/DRSpecialisationCompilerTest.class.st (hunk 2 above)
Comment (guillep) [reply Druid-Tests/DRSpecialisationCompilerTest.class.st:59]: Maybe for the test we can have a primitiveAdd without the pragma?

Comment (guillep) [reply Druid-Tests/DRSpecialisationCompilerTest.class.st:59]: Anyways, for a separate iteration

File: Druid/DRArrayPrimitiveSpecialisation.class.st (hunk 4 above)
Comment (guillep) [reply Druid/DRArrayPrimitiveSpecialisation.class.st:17]: I don't know, it's empty :)

File: Druid/DRSmallIntegerPrimitiveSpecialisation.class.st (hunk 10 above)
Comment (guillep) [reply Druid/DRSmallIntegerPrimitiveSpecialisation.class.st:16]: Well, there is customiseFor:, specializeFor: and this one

File: Druid/DRSCCP.class.st (hunk 11 above)
Comment (guillep) [reply Druid/DRSCCP.class.st:204]: ah yes

**Comment by guillep** [comment]
Now there are merge conflicts

**Comment by PalumboN** [comment]
Yes, I'll fix the conflicts and comments and then merge it :)

**Comment by PalumboN** [comment]
Well... The last two days I tried to fix the Cogit initialization for testing... but I couldn't... Too many initialization methods in many classes that use many global variables 😞

**Comment by PalumboN** [comment]
We need to rethink the specialization. 

Maybe would be easier to have a subclass of the _CompilerCompiler_ to perform a specific specialization and check which points need to be overridden.
Then we can move that code to a composition of objects to avoid a subclasses number explosion.

**Pull request closed without merging**